proposals/NNNN-add-result.md

```diff
@@ -0,0 +1,242 @@
+
+```swift
+func dataTask(with url: URL, completionHandler: @escaping (Data?, URLResponse?, Error?) -> Void) -> URLSessionDataTask
+```
+
+This can make it quite difficult to elegantly consume the results of these APIs:
+
+```swift
+URLSession.shared.dataTask(with: url) { (data, response, error) in
+    guard error != nil else { self.handleError(error!) }
+    
+    guard let data = data, let response = response else { return // Impossible? }
+    
+    handleResponse(response, data: data)
+}
+```
+
+While this code is only a few lines long, it exposes Swift's complete lack of automatic error handling for asynchronous APIs. Not only was the `error` forcibly unwrapped (or perhaps handled using a slightly less elegant `if` statement), but a possibly impossible scenario was created. What happens if `response` or `data` are `nil`? Is it even possible? It shouldn't be, but Swift currently lacks the ability to express this impossibility. Using `Result<Value, Error>` for the same scenario allows for much more elegant code:
+
+```swift
+URLSession.shared.dataTask(with: url) { (result: Result<(URLResponse?, Data?), Error>) in // Type added for illustration purposes.
+    switch result {
+    case .success(let response):
+        handleResponse(response.0, data: response.1)
+    case .failure(let error):
+        handleError(error)
+    }
+}
+```
+
+This API expresses exactly the intended result (either an error or data and response, never all or none) and allows them to be handled much more clearly. 
+
+#### More General Usage
+
+More generally, there are several scenarios in which `Result` can make error handling and the surrounding APIs more elegant.
+
+#### Delayed Handling
+
+There may be times where a developer wants to immediately execute a `throw`ing function but otherwise delay handling the error until a later time. Currently, if they wish to preserve the error, they must break the value and error apart, as is typically seen in completion handlers (see above). This is made even more obnoxious if the developer needs to store the results of more than one function.
+
+```swift
+// Properties 
+var configurationString: String?
+var configurationReadError: Error?
+
+do {
+    string = try String(contentsOfFile: configuration)
+} catch {
+    readError = error
+}
+
+// Sometime later...
+
+func doSomethingWithConfiguration() {
+    guard let configurationString = configurationString else { handle(configurationError!) }
+    
+    ... 
+}
+
+```
+
+This can be made somewhat cleaner by using a `(string: String?, error: Error?)` to store the result, but it would still have the same usage issues as in the asynchronous case. Using a `Result` is the appropriate answer here, especially with the convenience API for creating a result from a `throw`ing function.
+
+```swift
+let configuration = Result { try String(contentsOfFile: configuration) }
```

Comment:
I propose we add special syntax to do this, enabling us to avoid the `try` and extraneous closure.
```suggestion
let configuration = catch String(contentsOfFile: configuration)
```

Reply:
I've specifically kept any special syntax out of this proposal, both to keep the proposal and  implementation simpler and to keep the language and usage impacts more predictable. I think some special syntax might be called for, but I'd like to leave it as a separate proposal. But if it's important to get this proposal into review, I can certainly add it. I would require help to implement it however.